source/lac/petsc_solver.cc

```diff
@@ -230,6 +231,7 @@ namespace PETScWrappers
     KSPSetConvergenceTest (solver_data->ksp, &convergence_test,
                            reinterpret_cast<void *>(&solver_control),
                            PETSC_NULL);
+    AssertThrow (ierr == 0, ExcPETScError(ierr));
```

Comment:
You're double checking the same value `ierr` here because you forget to assign the result of `KSPSetConvergenceTest` to `ierr`.